### DISCUSSION TITLE
Asynchronous loading image lag problem

### DISCUSSION THREAD

**Discussion by Jason0247** [airsdk/Adobe-Runtime-Support#2856] [Q&A]
I need to load 1000 jpg images and save bitmapdata in an array. During this loading process, if my main program has animations or is manipulating other visual objects, it will cause the system to stall.
Is there any way that I can load sequence images without affecting front-end user interaction, so that it can be done in the background, like multithreading?
thanks
![image](https://github.com/airsdk/Adobe-Runtime-Support/assets/91530077/21810a93-e0bd-4ade-8471-1878e11d3590)

**Comment by M4rkY4** [comment]
Worker is exactly what you're asking https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/Worker.html
I've never used it myself, so can't comment on its stability or ease of use.

Alternatively, you can always give small pauses between calculations or calculate time passed each frame and if it exceeds 1/60 of a second (frame time) or even better - 1/100 of a second (to give system some time to resolve other tasks), you can postpone next calculation to the next frame.
If your images share size and encoding methods, you can simply calculate time on one image and adjust maxConcurrentLoads number based on Math.floor(1/60/oneImageTime)

**Reply by Jason0247** [reply]
I have learned about multithreading methods for implementation, but it seems that they are not applicable on IOS. I am not sure if the information I am looking for is incorrect. In fact, the current method of playing 1920 * 1080 sequence images is very efficient, but even if I cache the images in advance to an array and read the Bitmapdata in the array, the playback speed is not very ideal, especially on the iPad (2048 * 1536). I don't know if it is a rendering efficiency issue, even if I use multi threading, the rendering efficiency itself is like this.
Thank you.

**Reply by M4rkY4** [reply]
Hi,
I'm not sure if I'm experienced enough with this technical part, but there are couple of things I think it's important to point out:

iOS devices are known for having (relatively) small amount of RAM. Let's take 2000x1000 image (like rounded 1920x1080 but for easier calculations). So we have 2 million of pixels per image. Each pixel has 3 colors and let's say we have a byte for a color, so we have 2m * 3 = 6 Megabytes per uncompressed image.
I just encoded FullHD Jpeg and it has a size almost of 1MB, so compressed images are 6 times smaller and we have approx. 1GB of RAM for 1000 images. That way, as we store in RAM either uncompressed or compressed images, we expect a thousand of images to use at least 1 Gigabyte of RAM (6GB when uncompressed) which is a lot of RAM. I assume that's too much for your iPad as some RAM is mandatory to be used by the iPad OS, some by integrated GPU, some by your app or other apps on background that are actively moved to your tablet flash memory during the start of your high RAM usage.

So I would either advice you to decode images at runtime, like this:
1) decode first second of images (for ex. 30 images).
2) When playback starts, each frame you decode one image and forget one image that was shown at previous frame, so you always have 30 images stored in RAM to avoid some shattering

Or, as it seems that you're actually creating your own Video Player, but without those hardware and software optimizations on reading and drawing processes, actually to use an existing one. Just use video formats with high quality that are hardware decoded and optimized for this behavior. Some kind of VideoTexture or that StageVideo layer that is below DisplayList and Stage3D layers.
If you want only forward playback from the start - pick any codec with hardware decoding and good enough quality. If you want rewinding or many playback time switches, pick codecs carefully as you need an INT**RA**FRAME codec only, not INT**ER**FRAME !

**Reply by HaimZik** [reply]
First you should profile to see what actually causing the slowdown, you can use Scout to do it.
Instead  of loading 1000 images in one frame you can spread it, for example 100 images for 10 frames.
Measuring time has performance cost so don't do it for just one image load but over many, with the time measured
you can budget how many images you can load per frame. 

You can try using ImageDecodingPolicy.ON_LOAD to decode async.
https://help.adobe.com/en_US/as3/dev/WS52621785137562065a8e668112d98c8c4df-8000.html
var loaderContext:LoaderContext = new LoaderContext(); 
loaderContext.imageDecodingPolicy = ImageDecodingPolicy.ON_LOAD 
var loader:Loader = new Loader(); 
loader.load(new URLRequest("http://www.adobe.com/myimage.png"), loaderContext);

If you are uploading the images to a texture use uploadFromBitmapDataAsync function, if you use Starling it will use it if you pass a function to uploadBitmapData.

**Reply by Jason0247** [reply]
Yes, that's right. I won't load 1000 images into memory at the same time. The method I'm using now is to load one image and clear the memory usage of the previous system. This method is relatively smooth on the iPad (the new version of the iPad), but of course, it cannot be compared to the native efficiency.
In fact, if it is in the same direction, it would be convenient, but it requires forward and backward, so only sequence images can be used.
The above operations are limited to jpg only. If playing png sequences, the efficiency will decrease by at least 1 to 2 times.
